plugin/code_actions.py

```diff
@@ -63,52 +84,215 @@ def get_location_key(self, view: sublime.View, point: int) -> str:
+    If user has enabled a generic action that matches more specific session action
+    (for example user's a.b matching session's a.b.c), then the more specific (a.b.c) must be
+    returned as servers must receive only actions that they advertise support for.
+    """
+    matching_kinds = []
+    for session_action in session_actions:
+        enabled = False
+        action_parts = session_action.split('.')
+        for i in range(len(action_parts)):
+            current_part = '.'.join(action_parts[0:i + 1])
+            user_value = user_actions.get(current_part, None)
+            if isinstance(user_value, bool):
+                enabled = user_value
+        if enabled:
+            matching_kinds.append(session_action)
+    return matching_kinds
+
+
+class LspSaveCommand(sublime_plugin.TextCommand):
+    """
+    A command used as a substitute for native save command. Runs code actions before triggering the
+    native save command.
+    """
 
+    def __init__(self, view: sublime.View) -> None:
+        super().__init__(view)
+        self._task = None  # type: Optional[CodeActionOnSaveTask]
+
+    def run(self, edit: sublime.Edit) -> None:
+        if self._task:
+            self._task.cancel()
+        on_save_actions = self._get_code_actions_on_save(self.view.settings())
+        if on_save_actions:
+            self._task = CodeActionOnSaveTask(self.view, on_save_actions, self._trigger_native_save)
+            self._task.run()
+        else:
+            self._trigger_native_save()
+
+    def _get_code_actions_on_save(self, view_settings: Union[dict, sublime.Settings]) -> Dict[str, bool]:
+        view_code_actions = view_settings.get('lsp_code_actions_on_save') or {}
+        code_actions = settings.lsp_code_actions_on_save.copy()
+        code_actions.update(view_code_actions)
+        allowed_code_actions = dict()
+        for key, value in code_actions.items():
+            if key.startswith('source.'):
+                allowed_code_actions[key] = value
+        return allowed_code_actions
+
+    def _trigger_native_save(self) -> None:
+        self._task = None
+        self.view.run_command('save', {"async": True})
+
+
+class CodeActionOnSaveTask():
+    """
+    The main task that requests code actions from sessions and runs them.
+
+    The amount of time the task is allowed to run is defined by user-controlled setting. If the task
+    runs longer, the native save will be triggered before waiting for results.
+    """
+
+    def __init__(self, view: sublime.View, on_save_actions: Dict[str, bool], on_done: Callable[[], None]) -> None:
+        self._view = view
+        self._manager = windows.lookup(view.window())
+        self._on_save_actions = on_save_actions
+        self._on_done = on_done
+        self._completed = False
+        self._canceled = False
+        # TODO: For limiting number of iterations the task can run. Maybe not needed since there is a timeout?
+        self._iteration = 0
```

Comment:
I will remove it for now. I don't see a reason to optimize prematurely.

I've initially added it because `lsp-stylelint` will never shut up and will always return code action when asked. It's just that the code action will not include any text edits. Initially, I've handled that with this iteration limit but then I found a better solution (check if the view has changes after running code actions).